File: frontend/src/services/scraper.ts
import api from '../api';

export interface ScraperTarget {
  id: string;
  user_id: string;
  company_name: string;
  career_url?: string | null;
  keywords: string[];
  is_active: boolean;
  last_scraped?: string;
  // AI company research fields
  website?: string | null;
  jobs_url?: string | null;
  description?: string | null;
  industry?: string | null;
  headquarters?: string | null;
  company_size?: string | null;
  talking_points?: string[];
  research_status?: 'none' | 'pending' | 'completed' | 'failed';
  researched_at?: string | null;
  research_sources?: string[];
  created_at: string;
  updated_at: string;
}

export interface WatchCompanyPayload {
  company_name: string;
  keywords?: string[];
}

export interface ScrapedJob {
  id: string;
  target_id: string;
  user_id: string;
  title: string;
  url?: string;
  description_snippet?: string;
  matched_keywords: string[];
  is_new: boolean;
  discovered_at: string;
}

export interface PaginatedScrapedJobs {
  jobs: ScrapedJob[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

export interface ScraperTargetCreatePayload {
  company_name: string;
  career_url?: string;
  keywords: string[];
}

export interface ScraperTargetUpdatePayload {
  company_name?: string;
  career_url?: string;
  keywords?: string[];
  is_active?: boolean;
}

export interface GeneralScraperSource {
  id: string;
  name: string;
  url: string;
  source_type: string; // "rss", "preset_linkedin", "preset_arbeitnow"
  locations: string[];
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface GeneralScraperSourceCreatePayload {
  name: string;
  url: string;
  source_type: string;
  locations: string[];
}

export interface GeneralScraperSourceUpdatePayload {
  name?: string;
  url?: string;
  locations?: string[];
  is_active?: boolean;
}

export const scraperService = {
  async listTargets(): Promise<ScraperTarget[]> {
    const response = await api.get<ScraperTarget[]>('/scraper/targets');
    return response.data;
  },

  async addTarget(payload: ScraperTargetCreatePayload): Promise<ScraperTarget> {
    const response = await api.post<ScraperTarget>('/scraper/targets', payload);
    return response.data;
  },

  async watchCompany(payload: WatchCompanyPayload): Promise<ScraperTarget> {
    const response = await api.post<ScraperTarget>('/scraper/targets/watch', payload);
    return response.data;
  },

  async researchTarget(id: string): Promise<ScraperTarget> {
    const response = await api.post<ScraperTarget>(`/scraper/targets/${id}/research`);
    return response.data;
  },

  async updateTarget(id: string, payload: ScraperTargetUpdatePayload): Promise<ScraperTarget> {
    const response = await api.patch<ScraperTarget>(`/scraper/targets/${id}`, payload);
    return response.data;
  },

  async deleteTarget(id: string): Promise<void> {
    await api.delete(`/scraper/targets/${id}`);
  },

  async triggerScrape(id: string): Promise<ScrapedJob[]> {
    const response = await api.post<ScrapedJob[]>(`/scraper/targets/${id}/scrape`);
    return response.data;
  },

  async listDiscoveredJobs(
    targetId?: string,
    page: number = 1,
    limit: number = 10
  ): Promise<PaginatedScrapedJobs> {
    const params: Record<string, any> = { page, limit };
    if (targetId) params.target_id = targetId;
    const response = await api.get<PaginatedScrapedJobs>('/scraper/jobs', { params });
    return response.data;
  },

  async markJobRead(jobId: string): Promise<ScrapedJob> {
    const response = await api.patch<ScrapedJob>(`/scraper/jobs/${jobId}/read`);
    return response.data;
  },

  async listGeneralSources(): Promise<GeneralScraperSource[]> {
    const response = await api.get<GeneralScraperSource[]>('/scraper/general-sources');
    return response.data;
  },

  async addGeneralSource(payload: GeneralScraperSourceCreatePayload): Promise<GeneralScraperSource> {
    const response = await api.post<GeneralScraperSource>('/scraper/general-sources', payload);
    return response.data;
  },

  async updateGeneralSource(id: string, payload: GeneralScraperSourceUpdatePayload): Promise<GeneralScraperSource> {
    const response = await api.patch<GeneralScraperSource>(`/scraper/general-sources/${id}`, payload);
    return response.data;
  },

  async deleteGeneralSource(id: string): Promise<void> {
    await api.delete(`/scraper/general-sources/${id}`);
  },

  async triggerGeneralScrape(id: string): Promise<ScrapedJob[]> {
    const response = await api.post<ScrapedJob[]>(`/scraper/general-sources/${id}/scrape`);
    return response.data;
  },
};

File: backend/src/models/scraper.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class ScraperTarget(BaseModel):
    id: Optional[str] = Field(default=None, alias="_id")
    user_id: str
    company_name: str
    career_url: Optional[str] = None
    keywords: List[str] = []
    is_active: bool = True
    last_scraped: Optional[datetime] = None
    # AI company research fields
    website: Optional[str] = None
    jobs_url: Optional[str] = None
    description: Optional[str] = None
    industry: Optional[str] = None
    headquarters: Optional[str] = None
    company_size: Optional[str] = None
    talking_points: List[str] = []
    research_status: str = "none"  # none | pending | completed | failed
    researched_at: Optional[datetime] = None
    research_sources: List[str] = []
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True
        json_encoders = {datetime: lambda v: v.isoformat()}


class ScrapedJob(BaseModel):
    id: Optional[str] = Field(default=None, alias="_id")
    target_id: str
    user_id: str
    title: str
    url: Optional[str] = None
    description_snippet: Optional[str] = None
    matched_keywords: List[str] = []
    is_new: bool = True
    discovered_at: datetime = Field(default_factory=datetime.utcnow)
    # --- de-duplication + freshness (optional; older docs won't have them) ---
    dedup_key: Optional[str] = None          # normalized URL used to detect duplicates
    first_seen: Optional[datetime] = None    # first time this posting was discovered
    last_seen: Optional[datetime] = None     # most recent scrape that still saw it
    company: Optional[str] = None            # resolved company name, when known
    location: Optional[str] = None           # posting location, when known
    source: Optional[str] = None             # e.g. "greenhouse", "lever", "html", "rss"
    # --- years-of-experience requirement, extracted from the description via
    # a small regex dictionary (scraper_utils.extract_years_experience) ---
    years_experience_min: Optional[int] = None
    years_experience_max: Optional[int] = None
    years_experience_display: Optional[str] = None  # e.g. "3-5 years", "5+ years"

    class Config:
        populate_by_name = True
        json_encoders = {datetime: lambda v: v.isoformat()}


class ScraperTargetCreateRequest(BaseModel):
    company_name: str
    career_url: Optional[str] = None
    keywords: List[str] = []


class WatchCompanyRequest(BaseModel):
    """Add a company to the watchlist — AI research fills in whatever isn't given."""
    company_name: str
    keywords: List[str] = []
    preferred_provider: Optional[str] = None
    # Optional: if the user already knows the careers/jobs page, providing it
    # skips the AI having to search for and guess at the right URL.
    career_url: Optional[str] = None


class ScraperTargetUpdateRequest(BaseModel):
    company_name: Optional[str] = None
    career_url: Optional[str] = None
    keywords: Optional[List[str]] = None
    is_active: Optional[bool] = None


class ScraperTargetResponse(BaseModel):
    id: str
    user_id: str
    company_name: str
    career_url: Optional[str] = None
    keywords: List[str] = []
    is_active: bool
    last_scraped: Optional[datetime] = None
    website: Optional[str] = None
    jobs_url: Optional[str] = None
    description: Optional[str] = None
    industry: Optional[str] = None
    headquarters: Optional[str] = None
    company_size: Optional[str] = None
    talking_points: List[str] = []
    research_status: str = "none"
    researched_at: Optional[datetime] = None
    research_sources: List[str] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        populate_by_name = True


class ScrapedJobResponse(BaseModel):
    id: str
    target_id: str
    user_id: str
    title: str
    url: Optional[str] = None
    description_snippet: Optional[str] = None
    matched_keywords: List[str] = []
    is_new: bool
    discovered_at: datetime
    first_seen: Optional[datetime] = None
    last_seen: Optional[datetime] = None
    company: Optional[str] = None
    location: Optional[str] = None
    source: Optional[str] = None
    years_experience_min: Optional[int] = None
    years_experience_max: Optional[int] = None
    years_experience_display: Optional[str] = None

    class Config:
        populate_by_name = True


class PaginatedScrapedJobsResponse(BaseModel):
    jobs: List[ScrapedJobResponse]
    total: int
    page: int
    limit: int
    pages: int
